### ISSUE TITLE
Add wildcard imports

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#579]
Here's a proposed spec. Consider:

```
use io
```

Each `use` statement denotes a `Names -> Names`, and the `Names` just contain `Relation Name a` values that get transformed as follows: a wildcard import of `io` looks for any definition `d` prefixed by `io`, with a suffix of `suffix`, removes any entry previously assigned to `suffix`, and adds the entry `(suffix, d)` to the output relation.

Example, if `io.IO` is a type, then `use io` will bring it into scope with the local unqualified name `IO`.

In other words, it is exactly equivalent to `use io foo bar baz` for all symbols `foo`, `bar`, `baz`... that have FQNs that are prefixed by `io`.

**Comment by atacratic** [comment]
It will be good to have this, I think.  I guess the one high-level worry is that it breaks a kind of 'locality of understanding' that we had before: at the moment, I can look at any name, find the relevant use statement (or lack of) going up to the top of the function, and know the FQN I'm dealing with.  If instead I've got some wildcard imports, I can't do that any more - I don't know if the name is already an FQN, or else which of the wildcard imports it's coming from.  OK, I can fix that with a 'find' at the unison prompt, or maybe one-day my editing environment will tell me the FQN when I mouseover the name.  

I guess I'm assuming here that the pretty-printer emits wildcard imports - maybe that's not what you're thinking?  Maybe the wildcards are a convenience on initial input, when you know what name you're trying to refer to, but when you view the code later you get the more specific rendering.  I kind of like this approach.  I guess the question is whether all the `use Widget foo bar baz quux ...` stuff is too noisy.  I'm starting to get used to it (especially now it's in tasteful grey in `view` output... :smile:)

Some other questions...

### `hiding`
(I wrote this section assuming we'd be pretty-printing with wildcard imports.)  Consider the following.
```
-- Suppose we have names Widget.a, Widget.b ... Widget.j, and Widget.z.

-- nice pretty-print output
baz : Nat
baz = use Widget
      a + b + c + d + e + f + g + h + i + j

-- not nice pretty-print output
baz' : {State Nat} Nat
baz' = let
         z = get
         -- can't clobber local name so no `use Widget`
         z + Widget.a + Widget.b + Widget.c + Widget.d + Widget.e + Widget.f + Widget.g + Widget.h + Widget.i + Widget.j

```

How can we avoid the big discontinuity between `baz` and `baz'`, when there's a name in `Widget` we don't want to import?  Maybe we'd say (already say?) that local names can be introduced which clash with imported names, and that local ones win?
```
use Widget
z = get  -- Widget.z is now masked, so any reference has to be fully qualified
```
What about if the `z` is another import?
```
baz'' : {State Nat} Nat
baz'' = let
          use Bip z
          -- can't do `use Widget` here - or does the pretty-printer have to be so clever as to put it before the `use Bip z`?
          z * z + Widget.a + Widget.b + Widget.c + Widget.d + Widget.e + Widget.f + Widget.g + Widget.h + Widget.i + Widget.j

```

So the point is there's a name clash between two namespaces, at least one of which is being used intensively in your function, where you want to use the clashing name.  It's a shame if one of the wildcard imports has to be dropped.  I think this would be common enough that this would be a noticeable paper-cut, and it's the kind of thing that would really take the shine off the 'my code is slurped into Unison which stores it for me' approach for some people.  

Should we have `use Widget hiding z` or similar?

### order-sensitivity
This is a question about pre-existing behavior of `use`, but it's flagged in your proposal ('removes any previously assigned') so am worrying about it now.  Consider the following.
```
io.foo : Nat
io.foo = 3

my.foo : Nat
my.foo = 4

bar : Nat
bar = 
  use io foo
  use my foo
  foo

> bar 
--      | > bar  
--          ⧩
--          4
```
Because of the 'last `use` wins' behavior, the call to `foo` in `bar` has resolved to `my.foo`.  But the user could well be confused/mistaken - they might have meant `io.foo`.  Shouldn't we produce an error in this case?

### pretty-print strategy
If we do want the pretty-printer to emit wildcard imports, then how about this for a spec?  Let's say it emits `use Widget` whenever
 - it was about to emit `use Widget foo`
 - ~~there was going to be at least one other `use Widget bar` at or below that level~~ (this would need another annotation pass over the ABT during pretty-printing)
 - there is reference to some `Widget.baz` at or below that level (regardless of whether it was going to be given a non-wildcard `use` statement) 
 - `Widget` does not include any suffixes which are used at or below that level from other namespaces.
   - (This is in keeping with the spec for emitting use statements, that if there's any suffix being used from multiple namespaces, then all the usages are qualified.)

**Comment by pchiusano** [comment]
One thought - I’d leave prettyprinter alone, so you can be sloppy about using wildcards when entering the the code, but once the code is slurped up, you get nice, precise imports when the code is viewed.

**Comment by atacratic** [comment]
Nice, OK.  I think that vaporizes my other points, apart from the one about order sensitivity.  What do you think about that one?

**Comment by pchiusano** [comment]
I like idea of giving a parse error if there’s shadowing within the same imports list, though if there’s no wildcards I’m not sure how common that error will be. If there are wildcards, a simple convention like - always put your wildcard imports first might be fine. It’s useful to allow shadowing IMO, especially with wildcards involved, because alternative is needing to be very precise with hiding and whatnot, which gets quite tedious.

**Comment by pchiusano** [comment]
Okie dokie, added this. Syntax is just `use blah`, which is equivalent to `use blah x y z` for `blah.x`, `blah.y`, `blah.z` and any other names prefixed by `blah`.

<img width="1037" alt="Screen Shot 2019-07-17 at 11 28 57 PM" src="https://user-images.githubusercontent.com/11074/61427054-ae347d80-a8ea-11e9-9586-a114e600133c.png">

@atacratic I did not add any parse error on shadowing yet. I'd like to get some usage of this under our belt before deciding what (if any) kinds of shadowing we want to turn into an error.

**Comment by atacratic** [comment]
Nice. With any future work on parse errors for shadowing, it might be good to defer the error until there is actually some reference in the code whose resolution would depend on how the shadowing is resolved. Then it's less tedious, and you're only making the user pay the minimum price for writing non-ambiguous code, rather than the maximum.

I suspect that any time there's a suffix being used that could match multiple imported names, we should surface that to the user and make them give that name fully qualified. Certainly if one of the imports is wildcard. And I suspect also even if one of the imports is lexically closer to the usage, either in terms of scope or import order. People aren't going to be tracking where all the names are well enough in their heads to always get this right without some interaction with unison... even more so if there are import priority rules to remember/guess.

But yes, good to get some usage under our belts first rather than theorising.
